ignite/engine/engine.py

```diff
@@ -683,16 +686,28 @@ def switch_batch(engine):
 
         if self.state.max_epochs is None or self._is_done(self.state):
             # Create new state
-            if max_epochs is None:
-                max_epochs = 1
             if epoch_length is None:
                 epoch_length = self._get_data_length(data)
                 if epoch_length is not None and epoch_length < 1:
                     raise ValueError("Input data has zero size. Please provide non-empty data")
 
+            if epoch_length is not None:
```

Comment:
useless test

Reply:
I'm unable to calculate max_iters if epoch_length is None before the engine is running (a few test throws an error):
https://github.com/pytorch/ignite/blob/e6ef77d7bbeb2bde5edee7282f34efb958a124c4/tests/ignite/engine/test_engine.py#L638-L659

In those case I let max_iters stay as None as default and defined it when StopIterations calculates the epoch_length

Reply:
@sdesrozis `epoch_length` can still be None here if we deal with unknown size iterators. The check is used.
However, i'd prefer to do something else. Let's do a sort of "lazy" assignment for `max_epochs`, `max_iters`.
User passes:
1) `max_epochs=N`, `max_iters=None` and whatever `epoch_length` => everything should work as previously for N epochs
2) `max_epochs=None`, `max_iters=None` and whatever `epoch_length` => everything should work as previously for 1 epoch
3) `max_epochs=N`, `max_iters=M` and whatever `epoch_length` => raise an exception
4) `max_epochs=None`, `max_iters=M` and `epoch_length=K` => we can setup directly `max_epochs=ceil(max_iters//epoch_length)` and stop on `max_iters` iteration.
4) `max_epochs=None`, `max_iters=M` and `epoch_length=None` => we keep `max_epochs=None` and iterate until:
a) `max_iters` is reached
b) `StopIterations` is raised and we can calculate the `epoch_length` -> `max_epochs=ceil(max_iters//epoch_length)`.